gtm.cc

```diff
@@ -575,20 +577,28 @@ gtm_read()
     if (start_new || !trk_head) {
       trk_head = route_head_alloc();
       track_add_head(trk_head);
-      real_tr_count++;
-      if (!first_trk_head) {
-        first_trk_head = trk_head;
-      }
+      real_track_list.append(trk_head);
     }
     track_add_wpt(trk_head, wpt);
   }
 
   /* Tracklog styles */
-  trk_head = first_trk_head;
-  for (i = 0; i != ts_count && i != real_tr_count; i++) {
-    trk_head->rte_name = fread_string(file_in);
+  // TODO: The format document states there are ts_count tracklog style entries,
+  //       and tr_count tracklog entries.
+  //       Some tracklog entries may be contiuation entries, so we turn these into
+  //       real_track_list.size() <= tr_count tracks.
+  //       If ts_count != real_track_list.size() we don't know how to line up the tracklogs,
+  //       and the real tracks, with the tracklog styles.
+  if (ts_count != real_track_list.size()) {
+    warning(MYNAME ": The number of tracklog entries with the new flag set doesn't match the number of tracklog style entries.");
```

Comment:
If the condition is true then either
1) the input file hasn't been properly created, or
2) we don't understand the format.
The code below is now hardened against losing it's place in the file, it always reads ts_count * ( string + 12 bytes).  Since ts_count stored in the file I think we shouldn't read too many or too few bytes.

However, if the the condition is true, then we don't know how to line up the tracks and the track styles, and we are at risk of naming the tracks incorrectly.

With a warning we continue to read the file.  We could throw an error and give up, but the user might prefer something instead of nothing.  We could expand on the warning message and state "As a result the track names may be associated with the wrong tracks."

I think it is weird enough we want to know about it.

Suggestions?

Reply:
Convincing. We're potentially creating a slightly damaged file. We don't have a lot of GTM users represented, so we may not know.